src/s3ql/backends/b2/b2_backend.py

```diff
@@ -0,0 +1,807 @@
+import base64
+from dugong import (HTTPConnection, CaseInsensitiveDict, is_temp_network_error, BodyFollowing,
+                    ConnectionClosed, ConnectionTimedOut)
+from urllib.parse import urlparse
+import urllib
+import json
+import re
+import ssl
+import binascii
+from ast import literal_eval
+from itertools import count
+import io
+import os
+from shutil import copyfileobj
+
+from  typing import Dict, Tuple, Union, List
+
+from ..common import (AbstractBackend, get_ssl_context, retry, checksum_basic_mapping,
+                      CorruptedObjectError, NoSuchObject, DanglingStorageURLError)
+from ...inherit_docstrings import (ABCDocstMeta, copy_ancestor_docstring, prepend_ancestor_docstring)
+from ...logging import logging, QuietError
+from ... import BUFSIZE
+from .object_r import ObjectR
+from .object_w import ObjectW
+from .b2_error import B2Error, BadDigestError
+
+log = logging.getLogger(__name__)
+
+class B2Backend(AbstractBackend, metaclass=ABCDocstMeta):
+    '''A backend to store data in backblaze b2 cloud storage.
+    '''
+
+    known_options = { 'account-id', 'disable-versions', 'retry-on-cap-exceeded',
+                      'test-mode-fail-some-uploads', 'test-mode-expire-some-tokens', 'test-mode-force-cap-exceeded',
+                      'tcp-timeout' }
+
+    available_upload_url_infos = []
+
+    def __init__(self, options):
+        '''Initialize backend object
+        '''
+
+        super().__init__()
+
+        self.ssl_context = get_ssl_context(options.backend_options.get('ssl-ca-path', None))
+
+        self.options = options.backend_options
+
+        self.b2_application_key_id = options.backend_login
+        self.b2_application_key = options.backend_password
+
+        self.tcp_timeout = self.options.get('tcp-timeout', 20)
+
+        self.account_id = self.options.get('account-id', None)
+
+        self.disable_versions = self.options.get('disable-versions', False)
```

Comment:
The problem is that it is possible to set more than one versioning rule for a bucket. I think it would be best to leave it as it is. In my observations it has not happened that s3ql created a new version of the same block, and users should be able to set this correctly.

Reply:
ok